File: java-backend/src/main/java/com/forge/operations/api/PlatformAdminController.java
package com.forge.operations.api;

import com.forge.operations.service.PlatformAdminService;
import com.forge.shared.api.ApiEnvelope;
import com.forge.shared.security.AuthenticatedUser;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RequestPart;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api/platform-admin")
public class PlatformAdminController {

    private final PlatformAdminService service;

    public PlatformAdminController(PlatformAdminService service) {
        this.service = service;
    }

    private AuthenticatedUser caller(Authentication auth) {
        return (AuthenticatedUser) auth.getPrincipal();
    }

    /* ── Dashboard ──────────────────────────────────────────────────────── */
    @GetMapping("/dashboard")
    public ResponseEntity<?> dashboard(Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.getDashboardStats(caller(auth))));
    }

    /* ── Companies ──────────────────────────────────────────────────────── */
    @GetMapping("/companies")
    public ResponseEntity<?> listCompanies(Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.listCompanies(caller(auth))));
    }

    @GetMapping("/companies/{id}")
    public ResponseEntity<?> getCompany(
            @PathVariable UUID id, Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.getCompany(id, caller(auth))));
    }

    @PostMapping(value = "/companies", consumes = "multipart/form-data")
    public ResponseEntity<?> createCompany(
            @RequestParam Map<String, String> body,
            @RequestPart(value = "logo", required = false) MultipartFile logo,
            Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.createCompany(body, logo, caller(auth))));
    }

    @PutMapping(value = "/companies/{id}", consumes = "multipart/form-data")
    public ResponseEntity<?> updateCompany(
            @PathVariable UUID id,
            @RequestParam Map<String, String> body,
            @RequestPart(value = "logo", required = false) MultipartFile logo,
            Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.updateCompany(id, body, logo, caller(auth))));
    }

    @PostMapping("/companies/{id}/activate")
    public ResponseEntity<?> activateCompany(
            @PathVariable UUID id, Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.activateCompany(id, caller(auth))));
    }

    @PostMapping("/companies/{id}/deactivate")
    public ResponseEntity<?> deactivateCompany(
            @PathVariable UUID id, Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.deactivateCompany(id, caller(auth))));
    }

    @PostMapping("/companies/{id}/enter")
    public ResponseEntity<?> enterCompany(
            @PathVariable UUID id, Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.enterCompany(id, caller(auth))));
    }

    @DeleteMapping("/companies/{id}")
    public ResponseEntity<?> deleteCompany(
            @PathVariable UUID id, Authentication auth) {
        service.deleteCompany(id, caller(auth));
        return ResponseEntity.ok(ApiEnvelope.success(null));
    }

    @PostMapping("/companies/{id}/reset-password")
    public ResponseEntity<?> resetOwnerPassword(
            @PathVariable UUID id,
            @RequestBody Map<String, Object> body,
            Authentication auth) {
        String newPwd = body.get("password") == null ? null : body.get("password").toString();
        String email = service.resetOwnerPassword(id, newPwd, caller(auth));
        return ResponseEntity.ok(ApiEnvelope.success(Map.of("email", email, "force_reset", true)));
    }

    /* ── Company users ──────────────────────────────────────────────────── */
    @GetMapping("/companies/{id}/users")
    public ResponseEntity<?> listCompanyUsers(
            @PathVariable UUID id, Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.listCompanyUsers(id, caller(auth))));
    }

    @PostMapping("/companies/{id}/users")
    public ResponseEntity<?> createCompanyUser(
            @PathVariable UUID id,
            @RequestBody Map<String, Object> body,
            Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.createCompanyUser(id, body, caller(auth))));
    }

    @PutMapping("/companies/{id}/users/{userId}")
    public ResponseEntity<?> updateCompanyUser(
            @PathVariable UUID id,
            @PathVariable UUID userId,
            @RequestBody Map<String, Object> body,
            Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.updateCompanyUser(id, userId, body, caller(auth))));
    }

    @DeleteMapping("/companies/{id}/users/{userId}")
    public ResponseEntity<?> deleteCompanyUser(
            @PathVariable UUID id,
            @PathVariable UUID userId,
            Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.deleteCompanyUser(id, userId, caller(auth))));
    }

    @PostMapping("/companies/{id}/users/bulk-delete")
    public ResponseEntity<?> bulkDeleteCompanyUsers(
            @PathVariable UUID id,
            @RequestBody Map<String, Object> body,
            Authentication auth) {
        @SuppressWarnings("unchecked")
        List<String> rawIds = (List<String>) body.get("userIds");
        if (rawIds == null || rawIds.isEmpty()) {
            return ResponseEntity.badRequest().body(ApiEnvelope.error("No user IDs provided"));
        }
        List<UUID> userIds = rawIds.stream().map(UUID::fromString).toList();
        return ResponseEntity.ok(ApiEnvelope.success(service.bulkDeleteCompanyUsers(id, userIds, caller(auth))));
    }

    @GetMapping("/companies/{id}/activity")
    public ResponseEntity<?> getCompanyActivity(
            @PathVariable UUID id, Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.getCompanyActivity(id, caller(auth))));
    }

    /* ── Platform admin users (/users mirrors Node.js routes) ───────────── */
    @GetMapping("/users")
    public ResponseEntity<?> listPlatformAdmins(Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.listPlatformAdmins(caller(auth))));
    }

    @PostMapping("/users")
    public ResponseEntity<?> createPlatformAdmin(
            @RequestBody Map<String, Object> body, Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.createPlatformAdmin(body, caller(auth))));
    }

    @PutMapping("/users/{id}")
    public ResponseEntity<?> updatePlatformAdmin(
            @PathVariable UUID id,
            @RequestBody Map<String, Object> body,
            Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.updatePlatformAdmin(id, body, caller(auth))));
    }

    @DeleteMapping("/users/{id}")
    public ResponseEntity<?> deletePlatformAdmin(
            @PathVariable UUID id, Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.deletePlatformAdmin(id, caller(auth))));
    }

    @PostMapping("/users/bulk-delete")
    public ResponseEntity<?> bulkDeleteUsers(
            @RequestBody Map<String, Object> body, Authentication auth) {
        @SuppressWarnings("unchecked")
        List<String> rawIds = (List<String>) body.get("userIds");
        if (rawIds == null || rawIds.isEmpty()) {
            return ResponseEntity.badRequest().body(ApiEnvelope.error("No user IDs provided"));
        }
        List<UUID> userIds = rawIds.stream().map(UUID::fromString).toList();
        return ResponseEntity.ok(ApiEnvelope.success(service.bulkDeleteUsers(userIds, caller(auth))));
    }

    @PostMapping("/users/{userId}/reset-password")
    public ResponseEntity<?> resetUserPassword(
            @PathVariable UUID userId,
            @RequestBody Map<String, Object> body,
            Authentication auth) {
        String pwd = body.get("password") == null ? null : body.get("password").toString();
        if (pwd == null || pwd.length() < 6) {
            return ResponseEntity.badRequest().body(ApiEnvelope.error("Password must be at least 6 characters"));
        }
        return ResponseEntity.ok(ApiEnvelope.success(service.resetUserPassword(userId, pwd, caller(auth))));
    }

    /* ── Company owners ─────────────────────────────────────────────────── */
    @GetMapping("/company-owners")
    public ResponseEntity<?> getCompanyOwners(Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.getCompanyOwners(caller(auth))));
    }

    /* ── Subscription check ─────────────────────────────────────────────── */
    @PostMapping("/check-subscriptions")
    public ResponseEntity<?> checkSubscriptions(Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.checkSubscriptions(caller(auth))));
    }

    /* ── Roles overview ─────────────────────────────────────────────────── */
    @GetMapping("/roles-overview")
    public ResponseEntity<?> getRolesOverview(Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.getRolesOverview(caller(auth))));
    }

    /* ── Access control ─────────────────────────────────────────────────── */
    @GetMapping("/access-control/users")
    public ResponseEntity<?> accessControlUsers(Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.listAccessControlUsers(caller(auth))));
    }

    @GetMapping("/access-control/companies")
    public ResponseEntity<?> accessControlCompanies(Authentication auth) {
        return ResponseEntity.ok(ApiEnvelope.success(service.listAccessControlCompanies(caller(auth))));
    }

    /* ── User activity ──────────────────────────────────────────────────── */
    @GetMapping("/user-activity")
    public ResponseEntity<?> getUserActivity(
            @RequestParam("user_id") String userId, Authentication auth) {
        if (userId == null || userId.isBlank()) {
            return ResponseEntity.badRequest().body(ApiEnvelope.error("user_id query parameter is required"));
        }
        return ResponseEntity.ok(ApiEnvelope.success(service.getUserActivity(userId, caller(auth))));
    }
}

File: backend/src/controllers/platformAdminController.js
/**
 * Platform Admin Controller
 * Handles all platform-level admin operations.
 */
const fs = require('fs').promises;
const path = require('path');
const platformAdminService = require('../services/platformAdminService');

/** Read an uploaded logo file and return a base64 data-URI string */
async function _logoToBase64(file) {
  try {
    const buf = await fs.readFile(file.path);
    const ext = path.extname(file.originalname).toLowerCase();
    const mime = ext === '.png' ? 'image/png' : ext === '.svg' ? 'image/svg+xml' : ext === '.webp' ? 'image/webp' : 'image/jpeg';
    return `data:${mime};base64,${buf.toString('base64')}`;
  } catch (err) {
    console.warn('Logo base64 conversion failed:', err.message);
    return null;
  }
}

const platformAdminController = {
  // ── Dashboard ───────────────────────────────────────────────────────────
  getDashboardStats: async (req, res) => {
    try {
      const stats = await platformAdminService.getDashboardStats();
      res.json({ success: true, data: stats });
    } catch (err) {
      console.error('Platform admin dashboard error:', err);
      res.status(500).json({ success: false, message: err.message });
    }
  },

  // ── Company Management ──────────────────────────────────────────────────
  getAllCompanies: async (req, res) => {
    try {
      const companies = await platformAdminService.getAllCompanies(req.query);
      res.json({ success: true, data: companies });
    } catch (err) {
      console.error('Get companies error:', err);
      res.status(500).json({ success: false, message: err.message });
    }
  },

  getCompanyById: async (req, res) => {
    try {
      const company = await platformAdminService.getCompanyById(req.params.id);
      res.json({ success: true, data: company });
    } catch (err) {
      const status = err.message === 'Company not found' ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  createCompany: async (req, res) => {
    try {
      const data = { ...req.body };
      if (req.file) {
        data.logo_url = `/uploads/logos/${req.file.filename}`;
        data.logo_data = await _logoToBase64(req.file);
      }
      const result = await platformAdminService.createCompany(data);
      // Sync logo to R2 after create (now we have company id)
      if (req.file) {
        const r2 = require('../services/r2StorageService');
        r2.syncFileToR2(req.file.path, { companyId: result.id, section: 'logos' })
          .catch(e => console.error('[R2] Logo sync failed:', e.message));
      }
      res.status(201).json({ success: true, data: result });
    } catch (err) {
      const status = err.message.includes('already exists') ? 409 : 400;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  updateCompany: async (req, res) => {
    try {
      const data = { ...req.body };
      if (req.file) {
        data.logo_url = `/uploads/logos/${req.file.filename}`;
        data.logo_data = await _logoToBase64(req.file);
        // Sync to R2 cloud storage
        const r2 = require('../services/r2StorageService');
        r2.syncFileToR2(req.file.path, { companyId: req.params.id, section: 'logos' })
          .catch(e => console.error('[R2] Logo sync failed:', e.message));
      }
      const company = await platformAdminService.updateCompany(req.params.id, data);
      res.json({ success: true, data: company });
    } catch (err) {
      const status = err.message === 'Company not found' ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  activateCompany: async (req, res) => {
    try {
      const company = await platformAdminService.activateCompany(req.params.id);
      res.json({ success: true, data: company, message: 'Company activated' });
    } catch (err) {
      const status = err.message === 'Company not found' ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  deactivateCompany: async (req, res) => {
    try {
      const company = await platformAdminService.deactivateCompany(req.params.id);
      res.json({ success: true, data: company, message: 'Company deactivated' });
    } catch (err) {
      const status = err.message === 'Company not found' ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  resetCompanyAdminPassword: async (req, res) => {
    try {
      const { password } = req.body;
      if (!password || password.length < 6) {
        return res.status(400).json({ success: false, message: 'Password must be at least 6 characters' });
      }
      const result = await platformAdminService.resetCompanyAdminPassword(req.params.id, password);
      res.json({ success: true, data: result });
    } catch (err) {
      const status = err.message.includes('not found') ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  // ── Company User Management ─────────────────────────────────────────────
  getCompanyUsers: async (req, res) => {
    try {
      const result = await platformAdminService.getCompanyUsers(req.params.id);
      res.json({ success: true, data: result });
    } catch (err) {
      const status = err.message === 'Company not found' ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  addCompanyUser: async (req, res) => {
    try {
      const user = await platformAdminService.addCompanyUser(req.params.id, req.body);
      res.status(201).json({ success: true, data: user });
    } catch (err) {
      const status = err.message.includes('not found') ? 404 : err.message.includes('already exists') || err.message.includes('limit') ? 409 : 400;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  updateCompanyUser: async (req, res) => {
    try {
      const user = await platformAdminService.updateCompanyUser(req.params.id, req.params.userId, req.body);
      res.json({ success: true, data: user });
    } catch (err) {
      const status = err.message.includes('not found') ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  deleteCompanyUser: async (req, res) => {
    try {
      const result = await platformAdminService.deleteCompanyUser(req.params.id, req.params.userId);
      res.json({ success: true, data: result });
    } catch (err) {
      const status = err.message.includes('not found') ? 404 : err.message.includes('last admin') ? 400 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  bulkDeleteCompanyUsers: async (req, res) => {
    try {
      const { userIds } = req.body;
      if (!Array.isArray(userIds) || userIds.length === 0) {
        return res.status(400).json({ success: false, message: 'No user IDs provided' });
      }
      const result = await platformAdminService.bulkDeleteCompanyUsers(req.params.id, userIds);
      res.json({ success: true, data: result });
    } catch (err) {
      const status = err.message.includes('not found') ? 404 : err.message.includes('last admin') ? 400 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  // Direct bulk delete (any users, regardless of company)
  bulkDeleteUsers: async (req, res) => {
    try {
      const { userIds } = req.body;
      if (!Array.isArray(userIds) || userIds.length === 0) {
        return res.status(400).json({ success: false, message: 'No user IDs provided' });
      }
      const result = await platformAdminService.bulkDeleteUsers(userIds);
      res.json({ success: true, data: result });
    } catch (err) {
      const status = err.message.includes('not found') ? 404 : err.message.includes('Cannot delete') ? 400 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  getCompanyActivity: async (req, res) => {
    try {
      const activity = await platformAdminService.getCompanyActivity(req.params.id);
      res.json({ success: true, data: activity });
    } catch (err) {
      const status = err.message === 'Company not found' ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  enterCompany: async (req, res) => {
    try {
      const company = await platformAdminService.enterCompany(req.params.id);
      res.json({ success: true, data: company });
    } catch (err) {
      const status = err.message === 'Company not found' ? 404 : err.message === 'Company is inactive' ? 400 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  deleteCompany: async (req, res) => {
    try {
      const result = await platformAdminService.deleteCompany(req.params.id);
      res.json({ success: true, data: result });
    } catch (err) {
      const status = err.message === 'Company not found' ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  // ── Platform Admin Users ────────────────────────────────────────────────
  getPlatformAdmins: async (req, res) => {
    try {
      const admins = await platformAdminService.getPlatformAdmins();
      res.json({ success: true, data: admins });
    } catch (err) {
      res.status(500).json({ success: false, message: err.message });
    }
  },

  createPlatformAdmin: async (req, res) => {
    try {
      const admin = await platformAdminService.createPlatformAdmin(req.body);
      res.status(201).json({ success: true, data: admin });
    } catch (err) {
      const status = err.message.includes('already exists') ? 409 : 400;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  deletePlatformAdmin: async (req, res) => {
    try {
      const result = await platformAdminService.deletePlatformAdmin(req.params.id);
      res.json({ success: true, data: result });
    } catch (err) {
      const status = err.message.includes('not found') ? 404 : err.message.includes('last platform') ? 400 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  updatePlatformAdmin: async (req, res) => {
    try {
      const result = await platformAdminService.updatePlatformAdmin(req.params.id, req.body);
      res.json({ success: true, data: result });
    } catch (err) {
      const status = err.message.includes('not found') ? 404 : err.message.includes('already exists') ? 409 : 400;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  // ── Subscription ────────────────────────────────────────────────────────
  checkSubscriptions: async (req, res) => {
    try {
      const result = await platformAdminService.checkSubscriptions();
      res.json({ success: true, data: result });
    } catch (err) {
      res.status(500).json({ success: false, message: err.message });
    }
  },

  // ── Roles & Permissions Overview ────────────────────────────────────────
  getRolesOverview: async (req, res) => {
    try {
      const data = await platformAdminService.getRolesOverview();
      res.json({ success: true, data });
    } catch (err) {
      console.error('getRolesOverview error:', err);
      res.status(500).json({ success: false, message: err.message });
    }
  },

  // ── Company Owners ──────────────────────────────────────────────────────
  getCompanyOwners: async (req, res) => {
    try {
      const owners = await platformAdminService.getCompanyOwners();
      res.json({ success: true, data: owners });
    } catch (err) {
      res.status(500).json({ success: false, message: err.message });
    }
  },

  // ── Access Control: All Users with Filters ──────────────────────────────
  getAccessControlUsers: async (req, res) => {
    try {
      const users = await platformAdminService.getAccessControlUsers(req.query);
      res.json({ success: true, data: users });
    } catch (err) {
      res.status(500).json({ success: false, message: err.message });
    }
  },

  // ── Access Control: Companies with drill-down info ──────────────────────
  getAccessControlCompanies: async (req, res) => {
    try {
      const companies = await platformAdminService.getAccessControlCompanies();
      res.json({ success: true, data: companies });
    } catch (err) {
      res.status(500).json({ success: false, message: err.message });
    }
  },

  // ── Reset any user's password ───────────────────────────────────────────
  resetUserPassword: async (req, res) => {
    try {
      const { password } = req.body;
      if (!password || password.length < 6) {
        return res.status(400).json({ success: false, message: 'Password must be at least 6 characters' });
      }
      const result = await platformAdminService.resetUserPassword(req.params.userId, password);
      res.json({ success: true, data: result });
    } catch (err) {
      const status = err.message.includes('not found') ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },

  // ── User Activity Search ──────────────────────────────────────────────
  getUserActivity: async (req, res) => {
    try {
      const { user_id } = req.query;
      if (!user_id) {
        return res.status(400).json({ success: false, message: 'user_id query parameter is required' });
      }
      const result = await platformAdminService.getUserActivity(user_id);
      res.json({ success: true, data: result });
    } catch (err) {
      const status = err.message.includes('not found') ? 404 : 500;
      res.status(status).json({ success: false, message: err.message });
    }
  },
};

module.exports = platformAdminController;
